### ISSUE TITLE
Bug(typescript-estree): always parse mts/mjs as ESM for non-type-aware parsing

### ISSUE THREAD

**Issue by fisker** [typescript-eslint/typescript-eslint#9101]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Repro Code

```TypeScript
import { parse } from "@typescript-eslint/typescript-estree"

const ast = parse('await(1)', {
  project: false,
  filePath: 'foo.mts'
})

console.log(ast.body[0].expression)
```

### ESLint Config

```javascript
N/A
```

### tsconfig

```jsonc
None
```

### Expected Result

The expression should be `AwaitExpression` instead of `CallExpression`.

### Actual Result

As described above.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `7.8.0` |
| `TypeScript`                           | `5.4.5` |
| `node`                                 | `20` |

**Comment by JoshuaKGoldberg** [comment]
Editing the title because it took me a moment to figure this out. 😄

**Comment by goldentrash** [comment]
I tried to fix this issue! See #9121.
Below is a screenshot of the result.

- If the file is ESM
![image](https://github.com/typescript-eslint/typescript-eslint/assets/33625132/4640c677-f345-4eb8-bb74-e27ad7cbda16)

- If the file is not ESM
![image](https://github.com/typescript-eslint/typescript-eslint/assets/33625132/eee7f866-f00f-4f8e-b280-e52a590acca2)

**Comment by goldentrash** [comment]
In `/tyscript-estree/src/convert.ts/Converter/convertNode`, modifying `sourceType` when it is case `SyntaxKind.SourceFile` does not make sense, below is the detail.

I tried modifying the sourceFile in `convertNode` with `extension === ts.Extension.Mjs || extension === ts.Extension.Mts`, but only the AST root (SourceFile) is changed, not the multiple child Nodes.

It appears that **before `convertNode` is called, the ts AST is fully constructed** via `ts.createSourceFile` and then changed to the estree AST via `convertNode`.

![Screenshot of what I tried](https://github.com/typescript-eslint/typescript-eslint/assets/33625132/20545803-724c-4e20-aeb3-7da29411d5bd)

**Comment by goldentrash** [comment]
In `/tyscript-estree/src/create-program/createSourceFile.ts/createSourceFile`, adjusting `setExternalModuleIndicator`, parameter of the `ts.createSourceFile`, doesn't make sense, below is the detail.

First, I checked how typescript's default `setExternalModuleIndicator` works.

1. Traverses all expressions in the source code.
2. Find ESM expressions like `export`.
3. Returns that Node.
4. If no nodes were found, it returns the **metadata node** of the SourceFile.

The default `setExternalModuleIndicator` is inappropriate because the intent is to parse ESM files as ESM, even if they don't have an ESM expression like `export`.

However, I think it would be difficult to **create a proper metadata node and insert it into the `SourceFile`**, and **it would rely too strongly on an external module** (typescript).

**Comment by goldentrash** [comment]
In `/tyscript-estree/src/create-program/createSourceFile.ts/createSourceFile`, adjusting `indicator`, parameter of the `ts.createSourceFile`, doesn't work, below is the detail.

The Interface annotation says "Controls the format the file is detected as", so I was expecting that, but modifying the `indicator` didn't achieve the desired result. **It seems to be a parameter related to js Emit, not AST construction.**

![screenshot of what I tried](https://github.com/typescript-eslint/typescript-eslint/assets/33625132/4fc87c38-8cef-4e3e-b6e5-eca6a5fdef27)

**Comment by fisker** [comment]
@goldentrash  Thanks for looking into this.

I saw there is a `setExternalModuleIndicator` option in `typescript` package, is it going to help?

**Comment by goldentrash** [comment]
Hi @fisker!

I updated the PR to find a new way to do this **because my original suggestion was unacceptable**, and the new way I found was to modify `setExternalModuleIndicator`.

The reason I initially thought it was inappropriate to modify `setExternalModuleIndicator` was because I was concerned that it would be too difficult to design a new function based **on the behavior of the default**(but we didn't have to), and that it would be too dependent on external modules.

However, while analyzing the typesciprt code to find a new way, I found that it is very easy to create an `externalModuleIndicator` that **simply returns `true` unconditionally based on the file extension**, and I also found cases where the same thing was done in the typescript internal code.

See [PR's comment](https://github.com/typescript-eslint/typescript-eslint/pull/9121#issuecomment-2123847425) for more details.

**Comment by goldentrash** [comment]
> Hi @fisker!
> 
> I updated the PR to find a new way to do this **because my original suggestion was unacceptable**, and the new way I found was to modify `setExternalModuleIndicator`.
> 
> The reason I initially thought it was inappropriate to modify `setExternalModuleIndicator` was because I was concerned that it would be too difficult to design a new function based **on the behavior of the default**(but we didn't have to), and that it would be too dependent on external modules.
> 
> However, while analyzing the typesciprt code to find a new way, I found that it is very easy to create an `externalModuleIndicator` that **simply returns `true` unconditionally based on the file extension**, and I also found cases where the same thing was done in the typescript internal code.
> 
> See [PR's comment](https://github.com/typescript-eslint/typescript-eslint/pull/9121#issuecomment-2123847425) for more details.

Hi! **I'm always ready to update the PR** if I need to change my solution to fix this bug!
I'll keep an email alarm on for this issue :)
